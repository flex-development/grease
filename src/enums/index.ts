/**
 * @file Entry Point - Enums
 * @module grease/enums
 */

export { default as LogLevel } from './log-level'
export { default as ReleaseType } from './release-type'
