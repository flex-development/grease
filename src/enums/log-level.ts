/**
 * @file Enums - LogLevel
 * @module grease/enums/LogLevel
 */

import type * as consola from 'consola'

/**
 * Log levels.
 *
 * @see https://github.com/unjs/consola#log-level
 *
 * @enum {consola.LogLevel}
 */
const LogLevel = Object.freeze({
  DEBUG: 4,
  ERROR: 0,
  FATAL: 0,
  INFO: 3,
  LOG: 2,
  SILENT: Number.NEGATIVE_INFINITY,
  TRACE: 5,
  VERBOSE: Number.POSITIVE_INFINITY,
  WARN: 1
} as const)

export default LogLevel
