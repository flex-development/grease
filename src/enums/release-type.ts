/**
 * @file Enums - ReleaseType
 * @module grease/enums/ReleaseType
 */

import type semver from 'semver'

/**
 * Release types.
 *
 * @enum {semver.ReleaseType}
 */
enum ReleaseType {
  MAJOR = 'major',
  MINOR = 'minor',
  PATCH = 'patch',
  PREMAJOR = 'premajor',
  PREMINOR = 'preminor',
  PREPATCH = 'prepatch',
  PRERELEASE = 'prerelease'
}

export default ReleaseType
