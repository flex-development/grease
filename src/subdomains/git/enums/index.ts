/**
 * @file Entry Point - Enums
 * @module grease/git/enums
 */

export { default as RepositoryProvider } from './repository-provider'
