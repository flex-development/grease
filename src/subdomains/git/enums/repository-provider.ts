/**
 * @file Enums - RepositoryProvider
 * @module grease/git/enums/RepositoryProvider
 */

/**
 * Repository provider names.
 *
 * @enum {Lowercase<string>}
 */
enum RepositoryProvider {
  BITBUCKET = 'bitbucket',
  GITHUB = 'github',
  GITLAB = 'gitlab'
}

export default RepositoryProvider
