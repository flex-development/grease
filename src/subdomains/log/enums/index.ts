/**
 * @file Entry Point - Enums
 * @module grease/log/enums
 */

export { default as LogLevel } from './log-level'
export { default as UserLogLevel } from './log-level-user'
export { default as LogType } from './log-type'
