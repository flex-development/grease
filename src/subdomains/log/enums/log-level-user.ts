/**
 * @file Enums - UserLogLevel
 * @module grease/log/enums/UserLogLevel
 */

import type LogLevel from './log-level'

/**
 * User-friendly log levels.
 *
 * @enum {Lowercase<keyof typeof LogLevel>}
 */
enum UserLogLevel {
  DEBUG = 'debug',
  ERROR = 'error',
  FATAL = 'fatal',
  INFO = 'info',
  LOG = 'log',
  SILENT = 'silent',
  TRACE = 'trace',
  VERBOSE = 'verbose',
  WARN = 'warn'
}

export default UserLogLevel
