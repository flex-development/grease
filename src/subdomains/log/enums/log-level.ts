/**
 * @file Enums - LogLevel
 * @module grease/log/enums/LogLevel
 */

import type { Opaque } from '@flex-development/tutils'

/**
 * Log levels.
 *
 * @enum {number}
 */
enum LogLevel {
  DEBUG = 4,
  ERROR = 0,
  FATAL = -1,
  INFO = 3,
  LOG = 2,
  SILENT = <Opaque<number, '-INFINITY'>>Number.NEGATIVE_INFINITY,
  TRACE = 5,
  VERBOSE = <Opaque<number, 'INFINITY'>>Number.POSITIVE_INFINITY,
  WARN = 1
}

export default LogLevel
