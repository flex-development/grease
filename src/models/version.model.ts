/**
 * @file Models - Version
 * @module grease/models/Version
 */

import { ReleaseType } from '#src/enums'
import type { ReleaseVersion } from '#src/types'
import type { SemanticVersion } from '@flex-development/pkg-types'
import { cast, type Optional } from '@flex-development/tutils'
import semver, { SemVer } from 'semver'

/**
 * Semantic version data model.
 *
 * @see {@linkcode SemVer}
 *
 * @class
 * @extends {SemVer}
 */
class Version extends SemVer {
  /**
   * Semantic version string.
   *
   * @see {@linkcode SemanticVersion}
   *
   * @public
   * @override
   * @member {SemanticVersion} version
   */
  public override version: SemanticVersion

  /**
   * Create a new semantic version.
   *
   * @see {@linkcode SemVer}
   *
   * @param {SemVer | string} version - Semantic version instance or string
   */
  constructor(version: SemVer | string) {
    super(version, { includePrerelease: true, loose: false })
    this.version = cast(version.toString())
  }

  /**
   * Copy properties from the given semantic version.
   *
   * Range options will **not** be copied.
   *
   * @public
   *
   * @param {SemVer} version - Version to copy
   * @return {this} `this` version
   */
  public copy(version: SemVer): this {
    this.build = version.build
    this.major = version.major
    this.minor = version.minor
    this.patch = version.patch
    this.prerelease = version.prerelease
    this.raw = version.raw
    this.format()
    return this
  }

  /**
   * Bump `this` version.
   *
   * @see {@linkcode ReleaseVersion}
   *
   * @public
   * @override
   *
   * @param {ReleaseVersion} release - Release version or type
   * @param {Optional<string>?} preid - Prerelease identifier
   * @param {number?} prestart - Prerelease start range
   * @return {this} `this` version
   */
  public override inc(
    release: ReleaseVersion,
    preid: Optional<string> = undefined,
    prestart: number = 1
  ): this {
    /**
     * New version.
     *
     * @var {SemVer} version
     */
    let version: SemVer = new SemVer(this.version)

    // bump version
    switch (true) {
      case release === ReleaseType.MAJOR:
      case release === ReleaseType.MINOR:
      case release === ReleaseType.PATCH:
      case release === ReleaseType.PREMAJOR:
      case release === ReleaseType.PREMINOR:
      case release === ReleaseType.PREPATCH:
      case release === ReleaseType.PRERELEASE:
        // @ts-expect-error ts(2554) types are wrong
        version.inc(cast(release), preid, prestart)
        break
      case !!semver.valid(release):
        version = new SemVer(release)
        break
      default:
        throw new Error(`invalid increment argument: ${release}`, {
          cause: { release }
        })
    }

    return this.copy(version)
  }

  /**
   * Get a string representation of `this` semantic version.
   *
   * @see {@linkcode SemanticVersion}
   *
   * @public
   * @override
   *
   * @return {SemanticVersion} Stringified semantic version
   */
  public override toString(): SemanticVersion {
    return this.version
  }
}

export default Version
