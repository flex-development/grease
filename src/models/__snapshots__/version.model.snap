// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:models/Version > #toJSON > should return json-serializable semantic version 1`] = `
{
  "build": [
    "067c93cc68575a652d8609585bd64478c38dccfe",
  ],
  "major": 1,
  "minor": 2,
  "patch": 3,
  "prerelease": [
    "alpha",
    4,
  ],
  "version": "1.2.3-alpha.4+067c93cc68575a652d8609585bd64478c38dccfe",
}
`;
